### PR TITLE
Migrate E2E tests to V3 host

### PR DESCRIPTION
**Pull request by Hazhzeng**
## Description

1. Automatically fetch the latest v3 Functions Host for E2E test.
2. Compile v3 Functions Host on local environment.
3. Update on extensions.csproj to use the latest popular versions of extensions.

## Note
1. This setup.py only uses python builtin modules as it is a convention.
2. V3 function host no longer pass `sys` properties in bindings (checked by `git checkout dev; git diff v2.x`). Thus, I comment out the assertion in related eventhub tests.
image

---

### PR information

- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-functions-python-worker/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: tests/endtoend/test_eventhub_batch_functions.py [hunk 1]
```diff
@@ -126,8 +126,9 @@ def test_eventhub_multiple_with_metadata(self):
             self.assertIsNotNone(sys_props['Offset'])
 
             # Metadata check eventhub trigger name
-            self.assertEqual(metadata['sys']['MethodName'],
-                             'metadata_multiple')
+            # V3 Functions Host no longer pass 'sys' property to python sdk
```
Comment (Hazhzeng) [inline tests/endtoend/test_eventhub_batch_functions.py:129]: See PR description >> Note

File: setup.py [hunk 2]
```diff
@@ -191,36 +189,109 @@ def finalize_options(self):
...
+
+        try:
+            subprocess.run(
+                args=['dotnet', 'build', 'WebJobs.Script.sln', '-o', 'bin'],
+                check=True,
+                cwd=str(webhost_dir),
+                stdout=sys.stdout, stderr=sys.stderr)
+        except Exception:
+            print(f"Failed to compile webhost in {webhost_dir}. "
+                  ".NET Core SDK is required to build the solution",
```
Comment (vrdmr) [inline setup.py:290]: nit; Just add `https://aka.ms/dotnet-download` link to the error, with a recommendation of which version to install.

File: setup.py [hunk 3]
```diff
@@ -241,15 +312,24 @@ def _install_extensions(self):
         env['TERM'] = 'xterm'  # ncurses 6.1 workaround
         try:
             subprocess.run(
-                args=['dotnet', 'build', '-o', 'bin'], check=True,
+                args=['dotnet', 'build', '-o', '.'], check=True,
                 cwd=str(self.extensions_dir),
                 stdout=sys.stdout, stderr=sys.stderr, env=env)
         except Exception:
-            print(f"dotnet core SDK is required")
+            print(f"dotnet core SDK is required to build extensions")
```
Comment (vrdmr) [inline setup.py:319]: here as well, `https://aka.ms/dotnet-download`

**Review by vrdmr — APPROVED** [review]
Great work! :)
LGTM

File: tests/endtoend/test_eventhub_batch_functions.py [hunk 4]
```diff
@@ -126,8 +126,9 @@ def test_eventhub_multiple_with_metadata(self):
...
+            # self.assertEqual(metadata['sys']['MethodName'],
+            #                  'metadata_trigger')
```
Comment (vrdmr) [inline tests/endtoend/test_eventhub_batch_functions.py:131]: Do we need the commented code?

File: tests/endtoend/test_eventhub_functions.py [hunk 5]
```diff
@@ -96,4 +96,5 @@ def test_eventhub_trigger_with_metadata(self):
         self.assertIsNotNone(sys_props['Offset'])
 
         # Metadata check eventhub trigger name
-        self.assertEqual(metadata['sys']['MethodName'], 'metadata_trigger')
+        # V3 Functions Host no longer pass 'sys' property to python sdk
+        # self.assertEqual(metadata['sys']['MethodName'], 'metadata_trigger')
```
Comment (vrdmr) [inline tests/endtoend/test_eventhub_functions.py:100]: Same here?
Also, does this need to be documented for python's documentation - the change in the metadata?

File: tests/endtoend/test_eventhub_batch_functions.py (hunk 4 above)
Comment (Hazhzeng) [reply tests/endtoend/test_eventhub_batch_functions.py:131]: Yeah, that's right. I will clean it up.

File: tests/endtoend/test_eventhub_functions.py (hunk 5 above)
Comment (Hazhzeng) [reply tests/endtoend/test_eventhub_functions.py:100]: No, this is not documented. The customer will access the metadata field as a dictionary.

**Pull request merged**